File: src/calculator/button_grid.py
import gi
gi.require_version("Gtk", "4.0")
from gi.repository import Gtk

import os

from svg_button import SvgButton
from svg_logo import SvgLogo

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
ICONS_DIR = os.path.join(BASE_DIR, "..", "icons")

class ButtonGrid(Gtk.Grid):
    def __init__(self, calculator_window):
        super().__init__(row_spacing=5, column_spacing=5)
        self.set_row_homogeneous(True)
        self.set_column_homogeneous(True)
        self.calculator = calculator_window

        buttons = [
            ("7", self.on_button_clicked), ("8", self.on_button_clicked), ("9", self.on_button_clicked), ("/", self.on_button_clicked),
            ("4", self.on_button_clicked), ("5", self.on_button_clicked), ("6", self.on_button_clicked), ("*", self.on_button_clicked),
            ("1", self.on_button_clicked), ("2", self.on_button_clicked), ("3", self.on_button_clicked), ("-", self.on_button_clicked),
            (".", self.on_button_clicked), ("0", self.on_button_clicked), ("=", self.on_equal_clicked), ("+", self.on_button_clicked),
            ("C", self.on_clear_clicked),
        ]

        row = 0
        col = 0
        plus_row = plus_col = 0
        for idx, (label, callback) in enumerate(buttons):
            svg_path = self.get_icon_filename(label)
            button = SvgButton(svg_path, label, callback)
            button.set_hexpand(True)
            button.set_vexpand(True)
            button.set_halign(Gtk.Align.FILL)
            button.set_valign(Gtk.Align.FILL)
            self.attach(button, col, row, 1, 1)
            if label == "+":
                plus_row = row
                plus_col = col
            col += 1
            if col > 3:
                col = 0
                row += 1

        logo = SvgLogo(os.path.join(ICONS_DIR, "logo.svg"), size=100)
        logo.set_halign(Gtk.Align.CENTER)
        logo.set_valign(Gtk.Align.CENTER)
        logo.set_margin_start(10)
        logo.set_margin_end(10)
        logo.set_margin_top(5)
        logo.set_margin_bottom(5)
        self.attach(logo, plus_col, plus_row + 1, 1, 1)

    def get_icon_filename(self, label):
        label_map = {
            "+": "plus", "-": "minus", "*": "multiply", "/": "divide",
            "=": "equal", ".": "dot", "C": "clear"
        }
        name = label_map.get(label, label)
        return os.path.join(ICONS_DIR, f"{name}.svg")

    def on_button_clicked(self, button):
        current = self.calculator.entry.get_text()
        label = button.label_value
        self.calculator.entry.set_text(current + label)

    def on_equal_clicked(self, button):
        try:
            result = eval(self.calculator.entry.get_text())
            self.calculator.entry.set_text(str(result))
        except Exception:
            self.calculator.entry.set_text("Error")

    def on_clear_clicked(self, button):
        self.calculator.entry.set_text("")

File: src/widgets/button_grid.rs
use gtk4::prelude::*;
use gtk4::{Grid, Entry};
use crate::widgets::svg_button::SvgButton;
use crate::widgets::svg_logo::SvgLogo;
use crate::utils::safe_eval;

pub struct ButtonGrid {
    pub widget: Grid,
}

impl ButtonGrid {
    pub fn new(entry: &Entry) -> Self {
        let grid = Grid::builder()
            .row_spacing(5)
            .column_spacing(5)
            .row_homogeneous(true)
            .column_homogeneous(true)
            .build();

        let buttons = vec![
            ("7", 0, 0), ("8", 1, 0), ("9", 2, 0), ("/", 3, 0),
            ("4", 0, 1), ("5", 1, 1), ("6", 2, 1), ("*", 3, 1),
            ("1", 0, 2), ("2", 1, 2), ("3", 2, 2), ("-", 3, 2),
            (".", 0, 3), ("0", 1, 3), ("=", 2, 3), ("+", 3, 3),
            ("C", 0, 4), 
        ];

        let icons_dir = "python_original/src/icons"; 

        for (label, col, row) in buttons {
            let icon_name = match label {
                "+" => "plus", "-" => "minus", "*" => "multiply", "/" => "divide",
                "=" => "equal", "." => "dot", "C" => "clear",
                _ => label,
            };
            let svg_path = format!("{}/{}.svg", icons_dir, icon_name);
            
            let svg_btn = SvgButton::new(&svg_path, label);
            let btn_widget = svg_btn.widget; // The gtk::Button

            let entry_clone = entry.clone();
            let label_string = label.to_string();

            btn_widget.connect_clicked(move |_| {
                let current_text = entry_clone.text();
                
                if label_string == "C" {
                    entry_clone.set_text("");
                } else if label_string == "=" {
                    let expr = current_text.as_str();
                    match safe_eval(expr) {
                        Ok(val) => {
                             if val.fract() == 0.0 {
                                 entry_clone.set_text(&format!("{}", val as i64));
                             } else {
                                 entry_clone.set_text(&format!("{}", val));
                             }
                        },
                        Err(_) => entry_clone.set_text("Error"),
                    }
                } else {
                    let mut text = current_text.to_string();
                    if text == "Error" {
                         text = String::new();
                    }
                    text.push_str(&label_string);
                    entry_clone.set_text(&text);
                }
            });

            grid.attach(&btn_widget, col, row, 1, 1);
        }
        
        let logo_path = format!("{}/logo.svg", icons_dir);
        let logo = SvgLogo::new(&logo_path, 100);
        
        logo.widget.set_margin_start(10);
        logo.widget.set_margin_end(10);
        logo.widget.set_margin_top(5);
        logo.widget.set_margin_bottom(5);
        
        grid.attach(&logo.widget, 3, 4, 1, 1);

        Self { widget: grid }
    }
}
